Python/search/dfs.py

```diff
@@ -1,58 +1,111 @@
+        
+        # edges are added to the adjacency list
+        self.graph[vertex1].append(vertex2)
+        self.graph[vertex2].append(vertex1)
+
+    def dfs_helper(self,start):
+        # function to assist in calling dfs
+        # also prints the final dfs traversal
+
+        # checks if the start node is in the graph
+        if start not in self.graph:
+            print('\n'+str(start)+' not in graph.'+'\n')
+            return 1 # 1 is returned to avoid stopping the running of the program, allows the user to re-enter the start vertex
+        
+        # dfs implementation starts
+        print('\n'+'Implementing DFS...')
+        self.traversal = [] # the dfs traversal is added to the list in order
+        self.dfs(start)
+
+        # prints the dfs implementation neatly
+        for i in range(len(self.traversal)):
+            if i == (len(self.traversal)-1):
+                print(str(self.traversal[i]))
+            else:
+                print(str(self.traversal[i])+' -> ',end="")
+            
+    def dfs(self,node):
+        # dfs function
+        if node not in self.visited:            # checks if the node was already visited
+            self.traversal.append(node)         # if the node wasnt traversed already, it's appended to the traversal
+            self.visited.add(node)              # the node is marked visited
+            for neighbour in self.graph[node]:  # the neighbouring vertices of the node are explored one after the other     
+                self.dfs(neighbour)
+        
+# Code Implementation starts here
+
+# Graph instance is created
+G = Graph()
+
+# vertices are added to the graph
+for v in range(int(input('How many vertices are in the graph? '))):
+    G.add_vertex(str(input("Enter Vertex: ")))
+
+# edges are added to the graph
+for e in range(int(input('\n'+'How many edges are in the Graph? '))):
+    source,destination = input("Enter Source and Destination Vertices separated by a space: ").split()
+    val = G.add_edge(source,destination)
+
+    # If the incorrecct vertex was given, the values are taken again for this iteration
+    # This is added to not skip over a count of the edge
+    while val == 1:
+        source,destination = input("Enter Source and Destination Vertices separated by a space: ").split()
+        val = G.add_edge(source,destination)
+
+# start node for the dfs traversal is taken from the user
+start_node = input('\n'+"Enter the Start Node for traversal: ")
+
+# dfs is implemented
+ret = G.dfs_helper(start_node)
+
+# If the user entered a non-existent vertiex as start node, the following is implemented
+# 1 is returned if the vertex is non existent
+while ret==1:
+    start_node = input('\n'+"Enter the Start Node for traversal: ")
+    ret = G.dfs_helper(start_node)
+
+# To implement dfs for bi-directional graphs, replace G.add_edge in lines 88-96
+# with G.add_edge_bidirectional
```

Comment:
Always have a empty line at the end of your file.

Reply:
Add an example of how someone would run this code at the very end of this code file. 
Also add space and time complexities for the algorithm you have implemented.